File: 03_RW_no_starving/main.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define READER_COUNT 5
#define WRITER_COUNT 2
#define READER_LOOPS 3
#define WRITER_LOOPS 3

/* * Sinhronizacija:
 * - mutex:   štiti read_count
 * - rw_lock: štiti kritičnu sekciju (bilo jedan writer ili više readera)
 * - red:     sprečava writer starvation - kad writer stigne, zaključa "red" 
 * i novi čitaoci ne mogu da ga preskoče.
 */
int read_count = 0;
sem_t mutex;    /* Štiti read_count */
sem_t rw_lock;  /* Štiti pristup resursu */
sem_t red;      /* Red čekanja - sprečava starvation */

void *reader(void *arg) {
    int id = *(int *)arg;

    for (int i = 0; i < READER_LOOPS; i++) {
        /* Ulazak u red */
        sem_wait(&red);
        sem_post(&red);

        sem_wait(&mutex);
        read_count++;
        if (read_count == 1) {
            sem_wait(&rw_lock); /* Prvi čitalac zaključava bazu za pisce */
        }
        sem_post(&mutex);

        /* --- KRITIČNA SEKCIJA ZA ČITANJE --- */
        printf("[Reader %d] poceo citanje (Trenutno citaoca: %d)\n", id, read_count);
        fflush(stdout);
        
        usleep((50 + rand() % 100) * 1000); /* Simulacija čitanja */
        
        printf("[Reader %d] zavrsio citanje\n", id);
        fflush(stdout);
        /* ----------------------------------- */

        sem_wait(&mutex);
        read_count--;
        if (read_count == 0) {
            sem_post(&rw_lock); /* Poslednji čitalac otključava bazu za pisce */
        }
        sem_post(&mutex);

        usleep((100 + rand() % 300) * 1000);
    }
    return NULL;
}

void *writer(void *arg) {
    int id = *(int *)arg;

    for (int i = 0; i < WRITER_LOOPS; i++) {
        /* Pisac zaključava red kako bi sprečio nove čitaoce da ga preskoče */
        sem_wait(&red);
        sem_wait(&rw_lock); 
        sem_post(&red);    

        /* --- KRITIČNA SEKCIJA ZA PISANJE --- */
        printf("[Writer %d] počeo pisanje i dodavanje podataka...\n", id);
        fflush(stdout);
        
        usleep((80 + rand() % 150) * 1000); /* Simulacija pisanja */
        
        printf("[Writer %d] završio pisanje.\n", id);
        fflush(stdout);
        /* ----------------------------------- */

        sem_post(&rw_lock); 

        usleep((150 + rand() % 350) * 1000);
    }
    return NULL;
}

int main(void) {
    srand((unsigned)time(NULL));

    /* Inicijalizacija semafora na vrednost 1 */
    sem_init(&mutex,   0, 1);
    sem_init(&rw_lock, 0, 1);
    sem_init(&red,     0, 1);

    pthread_t readers[READER_COUNT];
    pthread_t writers[WRITER_COUNT];
    int reader_ids[READER_COUNT];
    int writer_ids[WRITER_COUNT];

    /* Kreiranje niti */
    for (int i = 0; i < READER_COUNT; i++) {
        reader_ids[i] = i + 1;
        pthread_create(&readers[i], NULL, reader, &reader_ids[i]);
    }
    for (int i = 0; i < WRITER_COUNT; i++) {
        writer_ids[i] = i + 1;
        pthread_create(&writers[i], NULL, writer, &writer_ids[i]);
    }

    /* Čekanje da se sve niti završe */
    for (int i = 0; i < READER_COUNT; i++) pthread_join(readers[i], NULL);
    for (int i = 0; i < WRITER_COUNT; i++) pthread_join(writers[i], NULL);

    /* Uništavanje semafora */
    sem_destroy(&mutex);
    sem_destroy(&rw_lock);
    sem_destroy(&red);

    printf("\n[USPEH] Svi readeri i writeri su uspesno zavrsili rad!\n");
    return 0;
}

File: 03_RW_no_starving/main.go
/*
 * Sinhronizacija:
 * - mu:      stiti read_count
 * - rwLock:  stiti pristup resursu (jedan writer ili vise readera)
 * - red:     sprecava writer starvation
 *
 * -----------------------------------------------------------
 * C vs Go:
 *   sem_wait(&s)  =  uzmi token  =  <-s         (prima iz kanala)
 *   sem_post(&s)  =  vrati token =  s <- struct{}{}  (salje u kanal)
 *
 *   Kanal kapaciteta 1, inicijalno sa jednom vrednoscu = otkljucan semafor
 * -----------------------------------------------------------
 */

package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	READER_COUNT = 5
	WRITER_COUNT = 2
	READER_LOOPS = 3
	WRITER_LOOPS = 3
)

var readCount int
var mu sync.Mutex

var rwLock = make(chan struct{}, 1)
var red = make(chan struct{}, 1)

func reader(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 0; i < READER_LOOPS; i++ {
		// Ulazak u red
		<-red             // sem_wait(&red)  - uzmi token
		red <- struct{}{} // sem_post(&red)  - odmah vrati (samo provera da writer ne ceka)

		mu.Lock()
		readCount++
		if readCount == 1 {
			<-rwLock // sem_wait(&rwLock) - prvi reader blokira writere
		}
		mu.Unlock()

		// KRITICNA SEKCIJA ZA CITANJE
		fmt.Printf("[Reader %d] poceo citanje (Trenutno citaoca: %d)\n", id, readCount)
		time.Sleep(time.Duration(50+rand.Intn(100)) * time.Millisecond)
		fmt.Printf("[Reader %d] zavrsio citanje\n", id)

		mu.Lock()
		readCount--
		if readCount == 0 {
			rwLock <- struct{}{} // sem_post(&rwLock) - poslednji reader pusta writere
		}
		mu.Unlock()

		time.Sleep(time.Duration(100+rand.Intn(300)) * time.Millisecond)
	}
}

func writer(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 0; i < WRITER_LOOPS; i++ {
		// Writer zauzima red - novi readeri moraju da cekaju
		<-red             // sem_wait(&red)   - zauzmi red
		<-rwLock          // sem_wait(&rwLock) - cekaj da readeri zavrse
		red <- struct{}{} // sem_post(&red)   - pusti red tek kad dobijemo rwLock

		// KRITICNA SEKCIJA ZA PISANJE
		fmt.Printf("[Writer %d] poceo pisanje...\n", id)
		time.Sleep(time.Duration(80+rand.Intn(150)) * time.Millisecond)
		fmt.Printf("[Writer %d] zavrsio pisanje.\n", id)

		rwLock <- struct{}{} // sem_post(&rwLock)

		time.Sleep(time.Duration(150+rand.Intn(350)) * time.Millisecond)
	}
}

func main() {
	// Inicijalizacija - stavi jedan token da budu "otkljucani"
	rwLock <- struct{}{}
	red <- struct{}{}

	var wg sync.WaitGroup

	for i := 1; i <= READER_COUNT; i++ {
		wg.Add(1)
		go reader(i, &wg)
	}
	for i := 1; i <= WRITER_COUNT; i++ {
		wg.Add(1)
		go writer(i, &wg)
	}

	wg.Wait()
	fmt.Println("\n[USPEH] Svi readeri i writeri su uspesno zavrsili rad!")
}
